fastestimator/dataset/interleave_dataset.py

```diff
@@ -0,0 +1,271 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+import random
+from typing import Any, Dict, Iterable, List, Mapping, Optional, Union
+
+from torch.utils.data import Dataset
+
+from fastestimator.dataset.dataset import DatasetSummary, FEDataset
+from fastestimator.util.base_util import to_list, warn
+from fastestimator.util.traceability_util import traceable
+
+
+@traceable()
+class InterleaveDataset(FEDataset):
+    """A Dataset class that can allow for a step-wise interleaving of multiple datasets. This will be useful for
+    training multi-task models when we vary dataset used on a per-step basis.
+
+    For example, given dataset `ds1`, and `ds2`, if we want to vary dataset between each step, the following 3 will
+    produce the same behavior: step0 getting samples from ds1, step1 getting samples from ds2, and repeat.
+    ```python
+    dataset = InterleaveDataset(datasets=[ds1, ds2])
+    dataset = InterleaveDataset(datasets=[ds1, ds2], pattern=[0, 1])
+    dataset = InterleaveDataset(datasets={"a": ds1, "b": ds2}, pattern=["a", "b"])
+    ```
+
+    To achieve a more complicated interleaving pattern, for example, 2 steps of `ds1` followed by 3 step of `ds2`:
+    ```python
+    dataset = InterleaveDataset(datasets=[ds1, ds2], pattern=[0, 0, 1, 1, 1])
+    dataset = InterleaveDataset(datasets={"a": ds1, "b": ds2}, pattern=["a", "a", "b", "b", "b"])
+    ```
+
+    When datasets are provided as dictionary, users can use its key as `ds_id` in other Pipeline Operators to apply
+    dataset-spefic operations (such as batching or preprocessing). For example, if we need `ds1` to go through
+    `Minmax` then form a batch of 32, and `ds2` to use `Zscore` then form a batch of 64:
+    ```python
+    dataset = InterleaveDataset(datasets={"a": ds1, "b": ds2}, pattern=["a", "b"])
+    pipeline = fe.Pipeline(train_data = dataset,
+                            ops=[Minmax(..., ds_id="a"),
+                                Zscore(..., ds_id="b"),
+                                Batch(batch_size=32, ds_id="a"),
+                                Batch(batch_size=64, ds_id="b")])
+    ```
+
+    Args:
+        datasets: List of datasets or a dictionary with key being the dataset name, value being the dataset.
+        pattern: The step-wise interleaving patterns. When datasets provided is a list, it requires list of integer
+            when `datasets` is list, requires a list of names when `datasets` is dictionary.
+    """
+    def __init__(self,
```

Comment:
Use typing @overload to help users catch the required dataset+pattern rule

Reply:
done